ARCHITECTURE.md

```diff
@@ -0,0 +1,95 @@
+# ink!ternals
+
+
+This document describes the architecture of ink!. The information
+here targets those who want to understand or modify the inner
+workings of this project.
+
+In general we treat documentation as a first-class citizen.
+All crates mentioned below should be documented really well.
+You can find the crate documentation on docs.rs or for our
+`master` branch under GitHub pages. So for `ink_lang` e.g.:
+
+* [https://docs.rs/ink_lang/latest/ink_lang](https://docs.rs/ink_lang/latest/ink_lang) (latest)
+* [https://paritytech.github.io/ink/ink_lang](https://paritytech.github.io/ink/ink_lang) (`master`)
+
+ink! is composed of a number of crates that are all found in the
+`crates/` folder. On a high-level those can be grouped as:
+
+* `lang`: The ink! language itself.
+* `allocator`: The allocator for dynamic heap space allocation in a contract.
+* `engine`: An off-chain testing engine, it simulates a blockchain
+   environment and allows mocking specified conditions.
+* `env`: Exposes environmental functions, like information about the caller
+   of a contract call, getting random entropy, or e.g. self-terminating the
+	 contract.
+* `metadata`: Describes how to call a contract (it's interface and the
+   types), its storage layout, etc..
+* `prelude`: Provides an interface to typical standard library types and
+  functionality (like `vec` or `string`). Since contracts are run in a
+	`no_std` environment we provide this crate as an entrypoint for accessing
+	functionality of the standard library.
+* `primitives`: Utilities that are used internally by multiple ink! crates.
+* `storage`: The collections that are available for contract developers
+  to put in a smart contracts storage.
+
+An important thing to note is that the crates are primarily run in
+a `no_std` environment, specifically in a Wasm sandbox execution
+environment on the blockchain itself.
```

Comment:
This is worth expanding on. It's an important bit of context to know that the code generated will not be run on a "normal" computer like most other Rust programs

Reply:
Can you think of other differences besides the allocator thing?

Reply:
I added some more text on this.